src/ros_filter.cpp

```diff
@@ -1901,11 +1905,16 @@ namespace RobotLocalization
         clearExpiredHistory(filter_.getLastMeasurementTime() - historyLength_);
       }
 
-      if (!loop_rate.sleep())
+      ros::Duration loop_elapsed_time = ros::Time::now() - loop_end_time;
+      ros::Duration sleep_time = loop_cycle_time - loop_elapsed_time;
+
+      if (!sleep_time.sleep())
       {
-        ROS_WARN_STREAM("Failed to meet update rate! Try decreasing the rate, limiting "
-                        "sensor output frequency, or limiting the number of sensors.");
+        ROS_WARN_STREAM("Failed to meet update rate! Took " << std::setprecision(20) << loop_elapsed_time.toSec() <<
+          " seconds. Try decreasing the rate, limiting sensor output frequency, or limiting the number of sensors.");
       }
+
```

Comment:
Another quirk about r_l is that we don't really know how long a given cycle will take between `spinOnce()` and integration of all the measurements. At the start of an iteration, we get N measurements, then process them, then publish, then sleep for the rest of the cycle. So we will reach this line at *exactly* the specified loop rate, but we will reach the line which [publishes](https://github.com/locusrobotics/robot_localization/pull/9/files#diff-dd5f9ec904b27c736579e4457a23a746R1876) our messages after we're done integrating measurements, and that time will vary. This means a `rostopic hz` on the r_l output odometry message will have some error around the specific publication rate.

We could use threads to get around this by integrating measurements as fast as we can, and just publishing the output at a fixed frequency.